### PR TITLE
redux-saga: Correct redux saga definitions related to saga helpers

### PR DESCRIPTION
**Pull request by jhsimms**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/redux-saga/redux-saga/tree/master/docs/api#takeeverypattern-saga-args -- See example code
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

Description:

The current type definitions don't match the actual library API.  The
"takeEvery" and "takeLatest" helpers accept a generator that takes
an action, not an optional "getState" function.

This is currently passing the tests because the tests use untyped
parameters, so it's being incorrectly inferred to be correct.
Typing the parameters in the test exposes the bug.

### CODE REVIEW

File: redux-saga/index.d.ts [hunk 1]
```diff
@@ -41,9 +41,11 @@ declare module 'redux-saga' {
     cancel(): void;
   }
 
-  export function takeEvery(pattern: Pattern, saga: Saga, ...args: any[]): { [Symbol.iterator](): IterableIterator<any> };
+  export type SagaHelperHandler = <T>(action: T) => Iterable<any>;
 
-  export function takeLatest(pattern: Pattern, saga: Saga, ...args: any[]): { [Symbol.iterator](): IterableIterator<any> };
+  export function takeEvery(pattern: Pattern, saga: SagaHelperHandler, ...args: any[]): { [Symbol.iterator](): IterableIterator<any> };
+
+  export function takeLatest(pattern: Pattern, saga: SagaHelperHandler, ...args: any[]): { [Symbol.iterator](): IterableIterator<any> };
```
Comment (DanielRosenwasser) [inline redux-saga/index.d.ts]: `IterableIterator<any>` should be fine for the return type.

File: redux-saga/redux-saga-tests.ts [hunk 2]
```diff
@@ -27,6 +25,11 @@ import {applyMiddleware, createStore} from 'redux';
 declare const delay: (ms: number) => Promise<any>;
 declare const fetchApi: (url: string) => Promise<any>;
 
+interface Action {
```
Comment (DanielRosenwasser) [inline redux-saga/redux-saga-tests.ts]: Should `Action` be generic on the payload?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by aikoven** [comment]
Redux-saga has bundled typings: https://github.com/redux-saga/redux-saga/blob/v0.14.3/index.d.ts

Also there's a PR (https://github.com/redux-saga/redux-saga/pull/740) with numerous fixes, which will hopefully be merged into the next minor version.

**Comment by DanielRosenwasser** [comment]
@aikoven @jhsimms are those pretty well-maintained? If so, we should remove this package and add it to the [notNeededPackages.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/notNeededPackages.json) file.

**Comment by aikoven** [comment]
@DanielRosenwasser Yep, they are actively maintained and I'm not planning to leave it any time soon. 

I also invite you to join the discussion on the mentioned PR.

**Comment by jhsimms** [comment]
Thanks @DanielRosenwasser @aikoven.  I submitted a follow-up PR #14244 to mark these as not needed.

**Comment by DanielRosenwasser** [comment]
Thanks guys!

**Pull request closed without merging**